File: app/src/main/java/com/slapshotapps/dragonshockey/activities/home/HomeFragment.java
package com.slapshotapps.dragonshockey.activities.home;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import android.os.Bundle;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.widget.Toast;
import androidx.fragment.app.Fragment;
import com.crashlytics.android.Crashlytics;
import com.google.firebase.analytics.FirebaseAnalytics;
import com.google.firebase.database.DatabaseException;
import com.google.firebase.database.FirebaseDatabase;
import com.slapshotapps.dragonshockey.Config;
import com.slapshotapps.dragonshockey.R;
import com.slapshotapps.dragonshockey.Utils.DragonsHockeyIntents;
import com.slapshotapps.dragonshockey.Utils.ProgressBarUtils;
import com.slapshotapps.dragonshockey.activities.MainActivity;
import com.slapshotapps.dragonshockey.databinding.ActivityHomeBinding;
import com.slapshotapps.dragonshockey.models.HomeContents;
import com.slapshotapps.dragonshockey.models.SeasonSchedule;
import com.slapshotapps.dragonshockey.observables.HomeScreenObserver;
import com.slapshotapps.dragonshockey.observables.ScheduleObserver;
import io.fabric.sdk.android.Fabric;
import java.util.Date;
import rx.Observable;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Action1;
import rx.functions.Func1;
import rx.schedulers.Schedulers;
import timber.log.Timber;

public class HomeFragment extends Fragment {

    private Subscription hockeyScheduleSubscription;

    private FirebaseDatabase firebaseDatabase;
    private FirebaseAnalytics firebaseAnalytics;
    private ActivityHomeBinding binding;


    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, R.layout.activity_home, container,false);

        return binding.getRoot();
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);

        //TODO
        //setSupportActionBar(binding.toolbar);
        //if (!Config.isRelease) {
        //    ActionBar actionBar = getSupportActionBar();
        //    actionBar.setTitle("CERT Dragons Hockey CERT");
        //}

        firebaseDatabase = FirebaseDatabase.getInstance();
        firebaseAnalytics = FirebaseAnalytics.getInstance(getContext());

        try {
            firebaseDatabase.setPersistenceEnabled(true);
        }
        catch (DatabaseException exception) {
            Timber.e("Unable to set persistance for Firebase");
        }
    }

    @Override
    public void onResume() {
        super.onResume();

        hockeyScheduleSubscription = ScheduleObserver.getHockeySchedule(firebaseDatabase)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .flatMap(new Func1<SeasonSchedule, Observable<HomeContents>>() {
                @Override
                public Observable<HomeContents> call(SeasonSchedule games) {
                    return HomeScreenObserver.getHomeScreen(firebaseDatabase, games, new Date());
                }
            })
            .subscribe(new Action1<HomeContents>() {
                @Override
                public void call(HomeContents homeContents) {
                    binding.setItem(new HomeScreenViewModel(homeContents));
                    //TODO
     //               ProgressBarUtils.hideProgressBar(binding.toolbarProgressBar);
                }
            }, new Action1<Throwable>() {
                @Override
                public void call(Throwable throwable) {
                    binding.setItem(new HomeScreenViewModel(null));
                    //TODO
                    //ProgressBarUtils.hideProgressBar(binding.toolbarProgressBar);
                    Toast.makeText(HomeFragment.this.getContext(), R.string.error_loading, Toast.LENGTH_LONG)
                        .show();
                }
            });

        //TODO
       // ProgressBarUtils.displayProgressBar(binding.toolbarProgressBar);
    }





    @Override
    public void onPause() {
        super.onPause();

        if (hockeyScheduleSubscription != null) {
            hockeyScheduleSubscription.unsubscribe();
            hockeyScheduleSubscription = null;
        }
    }

    //@Override
    //public void onViewSchedule() {
    //    Bundle bundle = new Bundle();
    //    bundle.putString(FirebaseAnalytics.Param.CONTENT_TYPE, "Schedule");
    //    bundle.putString(FirebaseAnalytics.Param.ITEM_ID, "900");
    //    firebaseAnalytics.logEvent(FirebaseAnalytics.Event.SELECT_CONTENT, bundle);
    //
    //    //startActivity(DragonsHockeyIntents.createScheduleIntent(this));
    //    startActivity(new Intent(getContext(), MainActivity.class));
    //}
    //
    //@Override
    //public void onViewStats() {
    //    Bundle bundle = new Bundle();
    //    bundle.putString(FirebaseAnalytics.Param.CONTENT_TYPE, "Stats");
    //    bundle.putString(FirebaseAnalytics.Param.ITEM_ID, "902");
    //    firebaseAnalytics.logEvent(FirebaseAnalytics.Event.SELECT_CONTENT, bundle);
    //
    //    startActivity(DragonsHockeyIntents.createStatsIntent(getContext()));
    //}
    //
    //@Override
    //public void onViewRoster() {
    //    Bundle bundle = new Bundle();
    //    bundle.putString(FirebaseAnalytics.Param.CONTENT_TYPE, "Roster");
    //    bundle.putString(FirebaseAnalytics.Param.ITEM_ID, "901");
    //    firebaseAnalytics.logEvent(FirebaseAnalytics.Event.SELECT_CONTENT, bundle);
    //
    //    startActivity(DragonsHockeyIntents.createRosterIntent(getContext()));
    //}
}

File: app/src/main/java/com/slapshotapps/dragonshockey/activities/home/HomeFragment.kt
package com.slapshotapps.dragonshockey.activities.home

import android.content.Context
import android.os.Bundle
import android.view.*
import android.widget.Toast
import androidx.core.view.MenuProvider
import androidx.lifecycle.Lifecycle
import com.google.firebase.database.DatabaseException
import com.google.firebase.database.FirebaseDatabase
import com.slapshotapps.dragonshockey.R
import com.slapshotapps.dragonshockey.utils.DragonsHockeyIntents
import com.slapshotapps.dragonshockey.activities.HockeyFragment
import com.slapshotapps.dragonshockey.databinding.ActivityHomeBinding
import com.slapshotapps.dragonshockey.observables.HomeScreenObserver
import com.slapshotapps.dragonshockey.observables.ScheduleObserver
import rx.Subscription
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import timber.log.Timber
import java.util.*


class HomeFragment : HockeyFragment() {

    private var hockeyScheduleSubscription: Subscription? = null
    private var _binding: ActivityHomeBinding? = null
    private val binding get() = _binding!!

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View {
        _binding = ActivityHomeBinding.inflate(inflater, container, false)

        val listener = actionBarListener
        listener?.setTitle(getString(R.string.dragons_hockey))

        setupMenu()

        return binding.root
    }

    private fun setupMenu(){
        activity?.addMenuProvider(object : MenuProvider {
            override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
                menuInflater.inflate(R.menu.menu_home, menu)
            }

            override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
                return when (menuItem.itemId) {
                    R.id.action_admin -> {
                        startActivity(DragonsHockeyIntents.createAdminAuthIntent(context))
                        true
                    }
                    else -> false
                }

            }
        }, viewLifecycleOwner,Lifecycle.State.RESUMED)
    }


    override fun onResumeWithCredentials() {
        updateData()
    }

    override fun noCredentialsOnResume() {
        val listener = actionBarListener
        listener?.hideProgressBar()
        Toast.makeText(this@HomeFragment.context, R.string.error_loading, Toast.LENGTH_LONG)
                .show()
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        firebaseDatabase = FirebaseDatabase.getInstance()


        try {
            firebaseDatabase!!.setPersistenceEnabled(true)
        } catch (exception: DatabaseException) {
            Timber.e("Unable to set persistance for Firebase")
        }
    }

    override fun onResume() {
        val listener = actionBarListener
        listener?.showProgressBar()
        super.onResume()
    }


    private fun updateData() {
        val listener = actionBarListener

        hockeyScheduleSubscription = ScheduleObserver.getHockeySchedule(firebaseDatabase)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .flatMap { games ->
                    HomeScreenObserver.getHomeScreen(firebaseDatabase, games, Date())
                }
                .subscribe({ homeContents ->
                    //binding.item = HomeScreenViewModel(homeContents)
                    listener?.hideProgressBar()
                }, {
                    //binding.item = HomeScreenViewModel(null)
                    listener?.hideProgressBar()
                    Toast.makeText(this@HomeFragment.context, R.string.error_loading, Toast.LENGTH_LONG)
                            .show()
                })
    }

    override fun onPause() {
        super.onPause()
        hockeyScheduleSubscription?.unsubscribe()
        hockeyScheduleSubscription = null
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}
